Add JSON API for mass deleting actors

MayberryZoom opened the pull request:
Adds a JSON API for mass deleting actors.

Per scenario/actor layer, actors can be mass deleted by the following conditions:
- all
- all within the specified groups
- all not within the specified groups

Additionally, exceptions can be made to not delete specific actors.

This is mainly useful for creating the lights out effect. This will be used for an RDV option to remove all lights, and can be useful for ROM hack creators to remove the lights in specific scenarios/rooms while selectively leaving specific actors untouched. There's probably some niche use case for the entities/sounds layers, but there was no reason to restrict this to only lights, so I left it open for any layer.

Also fixes a bug where when the door patcher renamed a shield, it would leave the name of the original shield in the actor groups. For the mass delete feature this created an issue where if a group contained a shield rando had modified, it would try to delete an actor that doesn't exist and error, so I decided it was best to fix it in this PR.

henriquegemignani commented on src/open_dread_rando/specific_patches/mass_delete_actors.py:
```diff
@@ -0,0 +1,66 @@
+
+    for reference in configuration["to_keep"]:
+        to_keep.add(ActorReferenceTuple(
+            reference["scenario"],
```
Converting this kind of dict into an `ActorReferenceTuple` should be a separate method.

You can then even change the entire thing to something like this:
```python
to_keep = {ActorReferenceTuple.from_dict(reference) for reference in configuration["to_keep"]}
```

henriquegemignani replied:
(or `to_keep  = set(map(ActorReferenceTuple.from_dict , configuration["to_keep"]))` but I don't think that's better)

henriquegemignani reviewed (changes requested):
Can you add unit tests for mass_delete_actors.py as a whole?

For the helper methods, should be doable if you mock `scenario.all_actors_in_actor_layer`/`scenario.get_actor_group` to return some data, and then you check to_remove/to_keep has what you expected. Meanwhile if you mock the helpers and `.remove_entity` it should be fairly simple to test `mass_delete_actors` as well

MayberryZoom commented on src/open_dread_rando/files/schema.json:
```diff
@@ -546,6 +546,104 @@
+                }
+            },
+            "required": ["to_remove"],
+            "default": {}
```
Because `to_remove` also has a default value, what ends up happening is:

1. `mass_delete_actors` gets defaulted to `{}`
1. `to_remove` and `to_keep` both get defaulted to `[]`

MayberryZoom replied:
Basically, this is working as intended

MayberryZoom commented on src/open_dread_rando/files/schema.json:
```diff
@@ -546,6 +546,104 @@
+                    "default": []
+                },
+                "to_keep": {
+                    "description": "A list of actors not to remove. Use this to keep specific actors from a provided scenario or actor group that",
```
honestly no idea what happened to the rest of that I remember typing it out lol

MayberryZoom commented:
I added the test, should get full coverage from that part. I also added a PatcherEditor fixture, which we can probably use in other places later on.

I did end up adding return values to the helper functions, but not because I wanted to use it for the test. I ran it a few times and for some reason it was slightly faster? I know for a fact Python passes sets via reference, not value, so I have no idea what would be so expensive about passing in the original set. To be fair, It could be something weird with my machine, and I only ran each method a few times then eyeballed it, so maybe it's not actually faster. It was in the range of two digits of milliseconds if I remember correctly though (something like 30ms), so that seems unlikely.

I definitely feel like this *should* be slower since it's just extra operations, but I guess I won't complain about a measured speed increase. And it also made the function arguments a little cleaner, which is probably more important in any case.
